mathbox: add axioms for finite unions.

benjub opened the pull request:
These additions came after looking at the recent efforts by @BTernaryTau to reduce usage of ax-pow and ax-un.  One can see that most of "finite set theory" can avoid them.  For instance, there are multiple examples where @BTernaryTau proved finite versions of existing theorems that avoid ax-pow and ax-un.  However, there are things that cannot be proved for bad reasons: I saw examples of theorems being proved for unordered pairs but whose analogues for unordered triples are not provable, which seems arbitrary.

The reason is that the axiom of unordered pair is formulated in theories which have at least some kind of unions, without which it is not very useful.  In the section comment in my mathbox, I give some "stupid" models that show that uselessness.

On the other hand, it looks like when renouncing to ax-un and ax-pow (and more powerful schemas like replacement), then the combination of the axiom of singleton and the axiom of binary union (together with nullary union, which is empty set), gives all finite unions, all tuples, and a reasonable framework.

Ping @digama0 since this touches axiomatization, and ping also @jkingdon since this seems to play well intuitionistically too.

jkingdon commented:
> On the other hand, it looks like when renouncing to ax-un and ax-pow (and more powerful schemas like replacement),

I'd be careful of saying "more powerful" without more elaboration. Power set is controversial because it is impredicative. I haven't seen similar concerns for the axiom of union or replacement (for example, CZF has the axiom of union and a strengthened form of replacement, but weakens power set). Nice find of https://mathoverflow.net/questions/81815/is-the-axiom-of-union-independent-of-the-rest-of-zf which is about the axiom of union and how it relates to other issues like how to state collection.

To be honest I don't really know whether ZF minus ax-pow has quite the same issues as CZF vs IZF, or other more familiar examples of predicative versus impredicative theories, although I have sort of assumed most of the issues carry over to those situations. I'm largely out of my depth in terms of going too deep, however.

benjub commented:
> > On the other hand, it looks like when renouncing to ax-un and ax-pow (and more powerful schemas like replacement),
> 
> I'd be careful of saying "more powerful" without more elaboration. Power set is controversial because it is impredicative. I haven't seen similar concerns for the axiom of union or replacement

That's actually because I was careful that I added the parenthesis "(and more powerful schemas like replacement)" that you quoted: notice how it restricts the rest of my sentence.  I added this precautionary restriction (which may or may not be necessary), since the article by Oman I cited shows that one can do a lot even without the axiom of union (though he seems to require powerset as well as replacement for most of his results).

Maybe I shouldn't have used the term "powerful" indeed.  Not because that axiom is or is not powerful, but because this is simplynot a clearly defined notion (it could be, in terms of ordinal analysis, for instance, but this is far beyond the present matters). Similarly, coining some axioms as "controversial" or a "source of concern" can be avoided. (On the other hand, replacement is indeed predicative contrary to powerset, and replacement is in CZF contrary to powerset.)

Anyway: what I wrote here and in the above post is pretty informal and no formal conclusion should be blindly taken from it.

In the same way that Oman's article is a nice study of the axiom of union, you have (in the framework of ZFC):
* for replacement: Kanamori, _In praise of replacement_
* for powerset: Gitman, Hamkins, Johnstone, _What is the theory ZFC without power set?_

jkingdon commented:
Totally fair, speaking of "predicative" versus "impredicative" is more promising than trying to speak of what is or is not controversial.

I gather those terms also are capable of having multiple meanings, but I should stop now before I get even farther from stuff I understand.
